### PR TITLE
Feature/new flags

### PR DESCRIPTION
**Pull request by davidrajones**
New set of flags for DA group.

### CODE REVIEW

**Comment by keiranmraine** [comment]
@davidrajones where can I find the specification document for the new flags?

File: config/flag.to.vcf.convert.ini [hunk 1]
```diff
@@ -139,4 +139,14 @@ info=1
...
+
+[cavemanMatchNormalProportionFlag]
+info=0
+id=CMNP
+description=Tumour sample mutant allele proportion - normal sample mutant allele proportion < maxCavemanMatchedNormalProportion (differs from MNP in using CaVEMan only seen reads as per VCF)
+
+[withinGapRangeFlag]
+info=0
+id=GAP
+description=If variant is within withinXBpOfDeletion of a deletion/gap seen in reads without the variant, the deletion is present in at least minGapPresentInReads percent of total reads and no variant showing reads show the deletion.
```
Comment (keiranmraine) [inline config/flag.to.vcf.convert.ini:152]: > If variant is within withinXBpOfDeletion of a deletion/gap seen in reads without the variant, the deletion is present in at least minGapPresentInReads percent of total reads and no variant showing reads show the deletion.

If variant is within withinXBpOfDeletion of an indel in reads without the variant and the indel is present in at least minGapPresentInReads percent of total reads and no variant reads have the indel.

File: lib/Sanger/CGP/CavemanPostProcessing.pm [hunk 2]
```diff
@@ -410,6 +430,26 @@ sub _get_soft_clip_count_from_cigar{
...
 
+sub _getDistanceFromGapInRead{
+  my ($cigar_array,$rdPosIndexOfInterest) = @_;
+  my $min_gap_dist = -1;
+  my $currentRp = 0;
+  foreach my $cigSect(@{$cigar_array}){
+    if($cigSect->[0] eq $MATCH_CIG || $cigSect->[0] eq $SKIP_CIG ||
+          $cigSect->[0] eq $INS_CIG || $cigSect->[0] eq $SOFT_CLIP_CIG){
+      $currentRp+=$cigSect->[1];
+    }elsif($cigSect->[0] eq $DEL_CIG){
```
Comment (keiranmraine) [inline lib/Sanger/CGP/CavemanPostProcessing.pm:441]: I'm a little confused here.  In the specification (rt721770) the language is a combination of:

* near in/deletions
* alignment gap
* deletion/gap

I'm not sure if you've clarified but I think this is for both insertions and deletions.

File: lib/Sanger/CGP/CavemanPostProcessing/Flagger.pm [hunk 3]
```diff
@@ -38,12 +39,28 @@ use Sanger::CGP::CavemanPostProcessing;
...
+const my $NEW_ALLELE_VCF_FORMAT => 'GT:FAZ:FCZ:FGZ:FTZ:RAZ:RCZ:RGZ:RTZ:PM';
+const my %OLD_ALLELE_VCF_FORMAT_INDEX_HASH => ('A' => 1, 'C' => 2, 'G' => 3, 'T' => 4, );
+const my %NEW_ALLELE_VCF_FORMAT_INDEX_HASH => ('A'=>[1,5], 'C' =>[2,6], 'G'=>[3,7], 'T'=>[4,8], );
+
+const my $MATCH_CIG => 'M';
+const my $SKIP_CIG => 'N';
+const my $INS_CIG => 'I';
+const my $DEL_CIG => 'D';
+const my $SOFT_CLIP_CIG => 'S';
+const my $HARD_CLIP_CIG => 'H';
```
Comment (keiranmraine) [inline lib/Sanger/CGP/CavemanPostProcessing/Flagger.pm:55]: These same constants are in `lib/Sanger/CGP/CavemanPostProcessing.pm` and should be moved out into a constants class.  Remember you can do:

```perl
use Sanger::CGP::CavemanPostProcessing::Const;
const my $C => 'Sanger::CGP::CavemanPostProcessing::Const';
# localise for speed (far faster than going through the sub everytime)
const my $MATCH_CIG => $C->MATCH_CIG;
const my $SKIP_CIG => $C->SKIP_CIG;
...
```

To do this you have to define them as subs but it is clean.

File: lib/Sanger/CGP/CavemanPostProcessing/Flagger.pm [hunk 4]
```diff
@@ -271,6 +294,26 @@ sub _tumFetch{
...
+      my $dp_end = $currentRp+$cigSect->[1];
+      my $tmp_dist = max(abs($rdPosIndexOfInterest-$dp_start),abs($dp_end-$rdPosIndexOfInterest));
+      if($tmp_dist < $min_gap_dist || $min_gap_dist == -1){
+        $min_gap_dist = $tmp_dist;
+      }
+    }
+  }
+  return $min_gap_dist;
+}
+
```
Comment (keiranmraine) [inline lib/Sanger/CGP/CavemanPostProcessing/Flagger.pm:316]: Duplication of code in `lib/Sanger/CGP/CavemanPostProcessing.pm`, is there something blocking moving this into a shared module?  It's already a function call so it needing to go to a different module isn't an overhead.

File: lib/Sanger/CGP/CavemanPostProcessor.pm [hunk 5]
```diff
@@ -489,6 +506,26 @@ sub _get_soft_clip_count_from_cigar{
...
+      my $dp_start = $currentRp+1;
+      my $dp_end = $currentRp+$cigSect->[1];
+      my $tmp_dist = max(abs($rdPosIndexOfInterest-$dp_start),abs($dp_end-$rdPosIndexOfInterest));
+      if($tmp_dist < $min_gap_dist || $min_gap_dist == -1){
+        $min_gap_dist = $tmp_dist;
+      }
+    }
+  }
+  return $min_gap_dist;
+}
```
Comment (keiranmraine) [inline lib/Sanger/CGP/CavemanPostProcessor.pm:527]: 3rd instance of this subroutine, it doesn't require `$self` so it needs collapsing to a common location.

**Review by keiranmraine — CHANGES_REQUESTED** [review]

File: config/flag.to.vcf.convert.ini [hunk 6]
```diff
@@ -139,4 +139,14 @@ info=1
...
 type=Float
 id=ASMD
-description=A soft flag median alignement score of reads showing the variant allele
\ No newline at end of file
+description=A soft flag median alignement score of reads showing the variant allele
+
+[cavemanMatchNormalProportionFlag]
+info=0
+id=CMNP
+description=Tumour sample mutant allele proportion - normal sample mutant allele proportion < maxCavemanMatchedNormalProportion (differs from MNP in using CaVEMan only seen reads as per VCF)
```
[the remark this answers is not in the document]
Comment (davidrajones) [reply config/flag.to.vcf.convert.ini:147]: Fixed in b9a1a99

File: lib/Sanger/CGP/CavemanPostProcessing.pm (hunk 2 above)
Comment (davidrajones) [reply lib/Sanger/CGP/CavemanPostProcessing.pm:441]: Question asked in rt721770

File: lib/Sanger/CGP/CavemanPostProcessing/Flagger.pm (hunk 4 above)
Comment (davidrajones) [reply lib/Sanger/CGP/CavemanPostProcessing/Flagger.pm:316]: Cleaned up code. Need to compare new and old and ensure same results then get rid of old code base as new one is far more flexible imho. Not for this feature though. 39ca2b5

File: lib/Sanger/CGP/CavemanPostProcessing/Flagger.pm (hunk 3 above)
Comment (davidrajones) [reply lib/Sanger/CGP/CavemanPostProcessing/Flagger.pm:55]: Need to compare new and old and ensure same results then get rid of old code base as new one is far more flexible imho. Not for this feature though. Agree constants need organising. Is that for this feature or a refactoring release?

Comment (keiranmraine) [reply lib/Sanger/CGP/CavemanPostProcessing/Flagger.pm:55]: WRT constants it depends how much code is affected, if they're only used in a couple of places they could be done now, completely your choice.  At least setup a milestone with tasks for anything we've discussed.  We can then present it as cleanup that needs to be done to make this more maintainable.

File: config/flag.to.vcf.convert.ini (hunk 1 above)
Comment (davidrajones) [reply config/flag.to.vcf.convert.ini]: Corrected in 9d4611f

File: lib/Sanger/CGP/CavemanPostProcessing/Flagger.pm (hunk 3 above)
Comment (davidrajones) [reply lib/Sanger/CGP/CavemanPostProcessing/Flagger.pm]: Removed 'new' unused flagging code to new feature branch.

File: lib/Sanger/CGP/CavemanPostProcessing.pm (hunk 2 above)
Comment (davidrajones) [reply lib/Sanger/CGP/CavemanPostProcessing.pm]: Insertions and deletions are included

**Comment by davidrajones** [comment]
Changes made according to suggestion (aside from constants)

File: lib/Sanger/CGP/CavemanPostProcessor.pm [hunk 7]
```diff
@@ -624,15 +656,21 @@ sub _callbackMatchedNormFetch{
       $str = -1;
     }
     my $cig_str = $algn->cigar_str; # expensive and reused so save to variable
-
     #Read population
     $this_read->{str} = $str;
-    $this_read->{qbase} = substr $a->qseq, $rdPosIndexOfInterest-1, 1;
+
+    #if($is_covered == 1){
```
Comment (keiranmraine) [inline lib/Sanger/CGP/CavemanPostProcessor.pm:662]: Commented `if` block, can this be removed?  Makes it look like it was in the middle of investigation.

**Review by keiranmraine — COMMENTED** [review]
Provided the commented `if` block isn't something that needs more than deletion this is ready.

**Pull request merged**